Woodoku/src/Entity/WoodokuBoard.py

```diff
@@ -25,89 +28,171 @@ def remove_blocks(self, blocks_coord: List[Tuple[int, int]]) -> None:
+        """Check if each coord has a block on it. If all of those coords are
+        occupied, return true. Otherwise, false.
+
+        Args:
+            blocks_coord (List[Tuple[int, int]]): list of coordinates to check
+
+        Raises:
+            CannotPlaceShapeError: raised when checking a invalid block
 
-        :param blocks_coord:
-        :return:
+        Returns:
+            bool: if any coordinates in `blocks_coord` is occupied
         """
-        pass
+        # TODO: raise the Error, according to docstring, the code below is just
+        # a placeholder
+        raise CannotPlaceShapeError(0, 0)
 
 
 class WoodokuBoard:
+    """A 9x9 Woodoku Board"""
+
     __scoreAgent: ScoreAgent
-    __representation: WoodokuBoardRepresentation
+    __representation: _WoodokuBoardRepresentation
 
     def __init__(self):
-        pass
+        self.__scoreAgent = ScoreAgent()
+        self.__representation = _WoodokuBoardRepresentation()
 
-    def __str__(self):
-        pass
-
-    def initialize(self):
-        pass
 
     def can_add_shape_to_board(self, shape: WoodokuShape) -> bool:
+        """Check if the woodoku shape can fit into the board. If all current
+        shapes are not be able to add, the game fails.
+
+        Exhaustively scans the board in a random fashion to average the complexity
+
+        Args:
+            shape (WoodokuShape): a woodoku shape for validation
+
+        Returns:
+            bool: if the `shape` fits the board
         """
-        Check if the woodoku shape can fit into the board. If all current shapes are not be able to add, the game fails.
-        :return: if can fit.
-        """
-        pass
+        # TODO: improve performance
+        rng = np.random.default_rng()
+        all_blocks = [(i, j) for i in range(9) for j in range(9)]
+        rng.shuffle(all_blocks)
+        print(all_blocks)
+
+        for row, col in all_blocks:
+            if self.can_add_shape_at_location(shape, x=row, y=col):
+                return True
+
+        return False
 
     def can_add_shape_at_location(self, shape: WoodokuShape, x: int, y: int) -> bool:
+        """Check if the woodoku `shape` can be placed into the board at location
+        `(x, y)` coordinates as the left upper corner of the shape.
+
+        Args:
+            shape (WoodokuShape): The shape needed to be checked
+            x (int): x coordinate
+            y (int): y coordinate
+
+        Returns:
+            bool: if `shape` can be added to `(x,y)`
         """
-        Check if the woodoku shape can be placed into the board at location (x, y) coordinates as the left upper corner
-        of the shape.
-        :param shape: the shape
-        :param x: x coordinate
-        :param y: y coordinate
-        :return: if can be placed at location.
-        """
-        pass
+        blocks = self.__map_shape_to_board_at(shape, x, y)
+        return not self.__representation.is_occupied(blocks)
```

Comment:
The definition for `_WoodokuBoardRepresentation.is_occupied` is unclear. If it returns true when **ALL** blocks are occpied, then the code here is incorrect.

Reply:
Will request and `_WoodokuBoardRepresentation.is_any_occupied` to resolve the problem